src/modules/navigator/navigator_main.cpp

```diff
@@ -652,18 +658,89 @@ Navigator::task_main()
 			navigation_mode_new = &_rcLoss;
 			break;
 
-		case vehicle_status_s::NAVIGATION_STATE_AUTO_RTL:
-			_pos_sp_triplet_published_invalid_once = false;
+		case vehicle_status_s::NAVIGATION_STATE_AUTO_RTL: {
+				_pos_sp_triplet_published_invalid_once = false;
 
-			// if RTL is set to use a mission landing and mission has a planned landing, then use MISSION
-			if (mission_landing_required() && on_mission_landing()) {
-				navigation_mode_new = &_mission;
+				const bool rtl_activated = _previous_vstatus.nav_state != vehicle_status_s::NAVIGATION_STATE_AUTO_RTL;
 
-			} else {
-				navigation_mode_new = &_rtl;
-			}
+				switch (rtl_type()) {
+				case RTL::RTL_LAND:
+					if (rtl_activated) {
+						mavlink_and_console_log_info(get_mavlink_log_pub(), "RTL LAND activated");
+					}
 
-			break;
+					// if RTL is set to use a mission landing and mission has a planned landing, then use MISSION
+					if (on_mission_landing() && !get_land_detected()->landed) {
+						navigation_mode_new = &_mission;
+
+					} else {
+						navigation_mode_new = &_rtl;
+					}
+
+					break;
+
+				case RTL::RTL_MISSION:
+					if (_mission.get_land_start_available() && !get_land_detected()->landed) {
+						// the mission contains a landing spot
+						if (_navigation_mode != &_mission) {
+							if (_navigation_mode == nullptr) {
+								// switching from an manual mode, go to landing if not already landing
+								if (!on_mission_landing()) {
+									start_mission_landing();
+								}
+
+							} else {
+								// switching from an auto mode, continue the mission from the closest item
+								_mission.set_closest_item_as_current();
+							}
+						}
+
+						if (rtl_activated) {
+							mavlink_and_console_log_info(get_mavlink_log_pub(), "RTL Mission activated, continue mission");
+						}
+
+						navigation_mode_new = &_mission;
+
+					} else {
+						// fly the mission in reverse if switching from a non-manual mode
```

Comment:
This chunk is hard to follow. Correct me if I'm wrong, but the "switch from nonmission" could be the default MissionReverse on_activation() behaviour. MissionReverse on_inactive() could simply disable if Mission proper is running.

Could you explain the check if already in RTL?

Reply:
- I could make it the default on_activation() behaviour but I prefer not to as the closest mission_item does not necessarily has to be the previous waypoint.

- I also want to reset it in any other mode as you do not know where you are when you changed from RTL to MANUAL and again to RTL

- If you are already in the navigation state AUTO_RTL, you are not in the RTL or manual mode, the vehicle is not landed and the reverse mission is not finished then the mission_reverse mode is activated. In any other case the regular RTL is executed.